jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java

```diff
@@ -58,10 +62,26 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
               AbsoluteUnixPath pathOnContainer =
                   AbsoluteUnixPath.get("/").resolve(extraDirectory.relativize(localPath));
               Instant modificationTime = modificationTimeProvider.apply(localPath, pathOnContainer);
-              FilePermissions permissions = extraDirectoryPermissions.get(pathOnContainer);
+              FilePermissions permissions =
+                  extraDirectoryPermissions.get(pathOnContainer.toString());
               if (permissions == null) {
+                // Check for matching globs
+                Path containerPath = Paths.get(pathOnContainer.toString());
+                for (Map.Entry<String, FilePermissions> entry :
+                    extraDirectoryPermissions.entrySet()) {
+                  PathMatcher pathMatcher =
+                      FileSystems.getDefault().getPathMatcher("glob:" + entry.getKey());
```

Comment:
This looks good. I just think we can instantiate `PathMatchers` before `new DirectoryWalker(...)`. That way, we don't have to instantiate them over and over for each file in `extraDirectory`. Then at least we won't instantiate them again within the same layer. This is a minor optimization, but I think it's worth, because

1. In practice, almost all the time, you will have some files whose permissions are not configured.
2. The optimization doesn't add new code but just amounts to moving the code outside `.walk()`.

We could even skip path strings that aren't glob, but I think this is unnecessary and rather adds more code.

WDYT?